src/cpp/rtps/builtin/discovery/endpoint/EDPServerListeners.cpp

```diff
@@ -78,47 +78,56 @@ void EDPServerPUBListener::onNewCacheChangeAdded(
     change->writer_info.previous = nullptr;
     change->writer_info.num_sent_submessages = 0;
 
-    // String to store the topic of the writer
-    std::string topic_name = "";
 
     // DATA(w) case: new writer or updated information about an existing writer
     if (change->kind == ALIVE)
     {
+        EndpointAddedCallback writer_added_callback =
+                std::bind(&EDPServerPUBListener::continue_with_writer, this, reader, change);
+
         // Note: add_writer_from_change() removes the change from the EDP publications' reader history, but it does not
         // return it to the pool
-        add_writer_from_change(reader, reader_history, change, sedp_, false);
+        add_writer_from_change(reader, reader_history, change, sedp_, false, writer_added_callback);
 
-        // Retrieve the topic after creating the WriterProxyData (in add_writer_from_change()). This way, not matter
-        // whether the DATA(w) is a new one or an update, the WriterProxyData exists, and so the topic can be retrieved
-        auto temp_writer_data = get_pdp()->get_temporary_writer_proxies_pool().get();
-        if (get_pdp()->lookupWriterProxyData(auxGUID, *temp_writer_data))
-        {
-            topic_name = temp_writer_data->topicName().to_string();
-        }
+        // Stop and wait for callback in case of TypeLookupService needed time to process the types
+        return;
     }
     // DATA(Uw) case
     else
     {
         EPROSIMA_LOG_INFO(RTPS_EDP_LISTENER, "Disposed Remote Writer, removing...");
 
-        // Retrieve the topic before removing the WriterProxyData. We need it to add the DATA(Uw) to the database
-        auto temp_writer_data = get_pdp()->get_temporary_writer_proxies_pool().get();
-        if (get_pdp()->lookupWriterProxyData(auxGUID, *temp_writer_data))
-        {
-            topic_name = temp_writer_data->topicName().to_string();
-        }
-        else
-        {
-            EPROSIMA_LOG_WARNING(RTPS_EDP_LISTENER, "Writer Proxy Data missing for change " << auxGUID);
-        }
-
-
         // Remove WriterProxy data information
         get_pdp()->removeWriterProxyData(auxGUID);
 
         // Removing change from history, not returning the change to the pool, since the ownership will be yielded to
         // the database
         reader_history->remove_change(reader_history->find_change(change), false);
+
+        // Continue without waiting
+        continue_with_writer(reader, change);
```

Comment:
This is not correct! The WriterProxyData has been removed in line 101 and then you call this function that looks for this WriterProxyData in line 124. Same consideration applies below to the ReaderProxyData.

Reply:
fixed in [391527d](https://github.com/eProsima/Fast-DDS/pull/4240/commits/391527df3729a32aadbeb5815d70fd172648d916)